wfe/wfe.go

```diff
@@ -1627,6 +1638,62 @@ func (wfe *WebFrontEndImpl) makeChallenges(authz *core.Authorization, request *h
 	return nil
 }
 
+// validateReplacementOrder performs several sanity checks on the order to
+// determine if the order is a replacement of an existing order. If the order is
+// not a replacement or is a valid replacement, no problem will be returned.
+// Otherwise the caller will receive a problem.
+func (wfe *WebFrontEndImpl) validateReplacementOrder(newOrder *core.Order) *acme.ProblemDetails {
+	if newOrder == nil {
+		return acme.InternalErrorProblem("Order is nil")
+	}
+
+	// Lock the order for reading
+	newOrder.RLock()
+	defer newOrder.RUnlock()
+
+	if newOrder.Replaces == "" {
+		wfe.log.Printf("Order %q is not a replacement\n", newOrder.ID)
+		return nil
+	}
+
+	certID, err := wfe.parseCertID(newOrder.Replaces)
+	if err != nil {
+		return acme.MalformedProblem(fmt.Sprintf("parsing ARI CertID failed: %s", err))
+	}
+
+	originalOrder, err := wfe.db.GetOrderByCertSerial(certID.SerialNumber)
+	if err != nil {
+		return acme.InternalErrorProblem(fmt.Sprintf("could not find an order for the given certificate: %s", err))
+	}
+
+	if originalOrder.Replaces != "" || !originalOrder.IsReplaced {
```

Comment:
If the original order itself replaced something else, that's fine. We want to encourage chains of replacement.